File: ruby/lib/runapi/nano_banana/resources/text_to_image.rb
# frozen_string_literal: true

module RunApi
  module NanoBanana
    module Resources
      # NanoBanana image generation resource.
      # Generate images from text prompts with various models and options.
      class TextToImage
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/nano_banana/text_to_image"

        RESPONSE_CLASS = Types::TextToImageResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedTextToImageResponse

        def initialize(http)
          @http = http
        end

        # Generate an image and wait until complete.
        #
        # @param params [Hash] generation parameters
        # @return [RunApi::NanoBanana::Types::CompletedTextToImageResponse] completed generation with image results
        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        # Create an image generation task.
        #
        # @param params [Hash] generation parameters
        # @return [RunApi::NanoBanana::Types::TextToImageResponse] task creation result with id
        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        # Get generation status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::NanoBanana::Types::TextToImageResponse] current generation status
        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["text-to-image"], params)

          raise Core::ValidationError, "prompt is required" unless param(params, :prompt)
        end
      end
    end
  end
end

File: python/src/runapi/nano_banana/resources/text_to_image.py
"""Nano Banana text-to-image resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource, ValidationError

from ..types import (
    ASPECT_RATIOS,
    GENERATION_MODELS,
    OUTPUT_FORMATS,
    OUTPUT_RESOLUTIONS,
    CompletedTextToImageResponse,
    TextToImageResponse,
)


class TextToImage(Resource):
    """Generate images from text prompts with various models and options."""

    ENDPOINT = "/api/v1/nano_banana/text_to_image"

    RESPONSE_CLASS = TextToImageResponse
    COMPLETED_RESPONSE_CLASS = CompletedTextToImageResponse

    def run(self, **params: Any) -> Any:
        """Create a text-to-image task and poll until it completes.

        Args:
            **params: Text-to-image parameters (model, prompt, ...).

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create a text-to-image task and return immediately with an ``id``.

        Args:
            **params: Text-to-image parameters (model, prompt, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of a text-to-image task.

        Args:
            id: The task id.

        Returns:
            The current status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        if not params.get("model"):
            raise ValidationError("model is required")
        if not params.get("prompt"):
            raise ValidationError("prompt is required")

        model = params.get("model")
        if model not in GENERATION_MODELS:
            raise ValidationError(f"Invalid model: {model}. Must be one of: {', '.join(GENERATION_MODELS)}")

        self._validate_optional(params, "aspect_ratio", ASPECT_RATIOS)
        self._validate_optional(params, "output_resolution", OUTPUT_RESOLUTIONS)
        self._validate_optional(params, "output_format", OUTPUT_FORMATS)
